notebooks/04_model_select_and_optimize/nni_ncf.ipynb

```diff
@@ -0,0 +1,762 @@
+    "    \"--surprise-reader\", SURPRISE_READER,\n",
+    "    \"--rating-metrics\", \" \".join(RATING_METRICS),\n",
+    "    \"--ranking-metrics\", \" \".join(RANKING_METRICS),\n",
+    "    \"--usercol\", USERCOL,\n",
+    "    \"--itemcol\", ITEMCOL,\n",
+    "    \"--k\", str(K),\n",
+    "    \"--random-state\", str(RANDOM_STATE),\n",
+    "    \"--epochs\", str(NUM_EPOCHS),\n",
+    "    \"--primary-metric\", PRIMARY_METRIC\n",
+    "])\n",
+    "\n",
+    "if BIASED:\n",
+    "    script_params += \" --biased\"\n",
+    "if VERBOSE:\n",
+    "    script_params += \" --verbose\"\n",
+    "if REMOVE_SEEN:\n",
+    "    script_params += \" --remove-seen\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We specify the search space for the NCF hyperparameters"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ncf_hyper_params = {\n",
+    "    'n_factors': {\"_type\": \"choice\", \"_value\": [2, 4, 8, 12]},\n",
+    "    'learning_rate': {\"_type\": \"uniform\", \"_value\": [1e-3, 1e-2]},\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with open(os.path.join(TMP_DIR, 'search_space_ncf.json'), 'w') as fp:\n",
+    "    json.dump(ncf_hyper_params, fp)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This config file follows the guidelines provided in [NNI Experiment Config instructions](https://github.com/microsoft/nni/blob/master/docs/en_US/Tutorial/ExperimentConfig.md).\n",
+    "\n",
+    "The options to pay attention to are\n",
+    "- The \"searchSpacePath\" which contains the space of hyperparameters we wanted to tune defined above\n",
+    "- The \"tuner\" which specifies the hyperparameter tuning algorithm that will sample from our search space and optimize our model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "config = {\n",
+    "    \"authorName\": \"default\",\n",
+    "    \"experimentName\": \"tensorflow_ncf\",\n",
+    "    \"trialConcurrency\": 8,\n",
+    "    \"maxExecDuration\": \"1h\",\n",
+    "    \"maxTrialNum\": MAX_TRIAL_NUM,\n",
+    "    \"trainingServicePlatform\": \"local\",\n",
+    "    # The path to Search Space\n",
+    "    \"searchSpacePath\": \"search_space_ncf.json\",\n",
+    "    \"useAnnotation\": False,\n",
+    "    \"logDir\": LOG_DIR,\n",
+    "    \"tuner\": {\n",
+    "        \"builtinTunerName\": \"TPE\",\n",
+    "        \"classArgs\": {\n",
+    "            #choice: maximize, minimize\n",
+    "            \"optimize_mode\": \"maximize\"\n",
+    "        }\n",
+    "    },\n",
+    "    # The path and the running command of trial\n",
+    "    \"trial\":  {\n",
+    "      \"command\": sys.prefix + \"/bin/python ncf_training.py\" + \" \" + script_params,\n",
```

Comment:
Use string formatting instead of + to concat
https://www.python.org/dev/peps/pep-3101/#format-strings

Also, any call out to spawn a seperate python process like this looks fishy. What are edge cases around doing this? If I have a conda env that I am running, will sys.prefix for sure get me the correct directory? The first thing thats jumps out to me is you may run python2 instead of python3.

Reply:
changed formatting and sys.executable